app/modules/zoo-api/zooapi.factory.js

```diff
@@ -4,10 +4,21 @@ require('./zooapi.module.js')
     .factory('zooAPI', zooAPI);
 
 var ApiClient = require('panoptes-client/lib/api-client');
-
+var OAuth = require('panoptes-client/lib/oauth')
 // @ngInject
-function zooAPI(zooAPIConfig) {
-
+function zooAPI($rootScope, zooAPIConfig) {
+    ApiClient.beforeEveryRequest = function() {
+        return OAuth.checkBearerToken()
+            .then(function (token) {
+                console.log('Token refreshed: ', token);
+            })
+            .catch(function (error) {
+                console.log('Failed to refresh token: ', error);
+                alert('Your session has finished. Please save your work and login again.')
+                $rootScope.$broadcast('auth:loginChange');
+                OAuth.signOut();
```

Comment:
Should this call `OAuth.signIn(url)` and sign them back in again with a new token?

Reply:
I should probably clarify here. They aren't signed out yet here, but their token is about to expire. Calling `OAuth.signIn` will redirect the browser to panoptes.zooniverse.org, then back to SW with a refreshed token for another 2 hours.